Lib9c/Action/CombinationEquipment.cs

```diff
@@ -20,262 +17,358 @@
-            result.id = mail.id;
-            avatarState.UpdateV3(mail);
-            avatarState.questList.UpdateCombinationEquipmentQuest(RecipeId);
-            avatarState.UpdateFromCombination(equipment);
-            avatarState.UpdateQuestRewards(materialSheet);
+            slotState.Update(attachmentResult, context.BlockIndex, endBlockIndex);
+            // ~Update Slot
+
+            // Create Mail
+            var mail = new CombinationMail(
+                attachmentResult,
+                context.BlockIndex,
+                mailId,
+                endBlockIndex);
+            avatarState.UpdateV4(mail, context.BlockIndex);
+            // ~Create Mail
+
             return states
-                .SetState(AvatarAddress, avatarState.SerializeV2())
+                .SetState(avatarAddress, avatarState.SerializeV2())
                 .SetState(inventoryAddress, avatarState.inventory.Serialize())
                 .SetState(worldInformationAddress, avatarState.worldInformation.Serialize())
                 .SetState(questListAddress, avatarState.questList.Serialize())
                 .SetState(slotAddress, slotState.Serialize())
-                .SetState(ctx.Signer, agentState.Serialize());
+                .SetState(context.Signer, agentState.Serialize());
         }
 
-        protected override IImmutableDictionary<string, IValue> PlainValueInternal =>
-            new Dictionary<string, IValue>
-            {
-                ["avatarAddress"] = AvatarAddress.Serialize(),
-                ["recipeId"] = RecipeId.Serialize(),
-                ["subRecipeId"] = SubRecipeId.Serialize(),
-                ["slotIndex"] = SlotIndex.Serialize(),
-            }.ToImmutableDictionary();
-
-        protected override void LoadPlainValueInternal(
-            IImmutableDictionary<string, IValue> plainValue)
+        public static void AddAndUnlockOption(
+            AgentState agentState,
+            Equipment equipment,
+            IRandom random,
+            EquipmentItemSubRecipeSheetV2.Row subRecipe,
+            EquipmentItemOptionSheet optionSheet,
+            SkillSheet skillSheet
+        )
         {
-            AvatarAddress = plainValue["avatarAddress"].ToAddress();
-            RecipeId = plainValue["recipeId"].ToInteger();
-            SubRecipeId = plainValue["subRecipeId"].ToNullableInteger();
-            SlotIndex = plainValue["slotIndex"].ToInteger();
-        }
+            foreach (var optionInfo in subRecipe.Options.OrderByDescending(e => e.Ratio))
+            {
+                if (!optionSheet.TryGetValue(optionInfo.Id, out var optionRow))
+                {
+                    continue;
+                }
 
-        public static StatMap GetStat(EquipmentItemOptionSheet.Row row, IRandom random)
-        {
-            var value = random.Next(row.StatMin, row.StatMax + 1);
-            return new StatMap(row.StatType, value);
-        }
+                var value = random.Next(1, 100001) * 0.00001m;
```

Comment:
I think it would be better to use WeightedSelector.

Reply:
Why you think so?